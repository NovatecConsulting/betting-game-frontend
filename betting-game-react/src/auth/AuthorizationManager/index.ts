export { default as AuthorizationCodeFlowPKCEManager } from "./AuthorizationManager";
