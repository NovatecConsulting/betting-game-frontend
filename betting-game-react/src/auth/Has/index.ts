export { default as Has } from "./Has";
