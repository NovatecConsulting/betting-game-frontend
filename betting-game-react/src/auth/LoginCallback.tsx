import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import { AuthorizationManager } from ".";
import { ROUTE_INDEX } from "../routes";

const LoginCallback: React.FC = () => {
    const [loggedIn, setLoggedIn] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);
    React.useEffect(() => {
        const doLoginAndRedirect = async () => {
            try {
                await AuthorizationManager.loginCallback();
                setLoggedIn(true);
            } catch (e) {
                setError(e.message);
            }
        };
        doLoginAndRedirect();
    }, []);
    // TODO: use Loading Component
    return error ? (
        <div>{error}</div>
    ) : loggedIn ? (
        <Redirect to={ROUTE_INDEX} />
    ) : (
        <div>Loading...</div>
    );
};

export default LoginCallback;
