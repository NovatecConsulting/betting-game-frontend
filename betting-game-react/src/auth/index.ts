export { default as AuthContext, AuthorizationManager } from "./AuthContext";
export { default as LoginCallback } from "./LoginCallback";
export * from "./permissions";
