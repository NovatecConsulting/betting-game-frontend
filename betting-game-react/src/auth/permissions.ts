export const READ_TEAMS = "read:teams";
export const READ_MATCHDAYS = "read:matchdays";
