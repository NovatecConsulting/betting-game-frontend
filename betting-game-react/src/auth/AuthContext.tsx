import React from "react";
import { AuthorizationCodeFlowPKCEManager } from "./AuthorizationManager";

export let AuthorizationManager: AuthorizationCodeFlowPKCEManager;

const AuthContext: React.FC = ({ children }) => {
    // TODO: use config
    AuthorizationManager = new AuthorizationCodeFlowPKCEManager({
        authority: "https://betting-game.eu.auth0.com",
        client_id: "rqlyAphr2es3RUjYlPAEkf2VCldAR5wL",
        redirect_uri: "http://localhost:3000/login-callback",
        response_type: "code",
        response_mode: "query", // TODO: in the future hopefully "fragment" is supported
        scope: "openid profile email offline_access",
        post_logout_redirect_uri: "http://localhost:3000/",
        revokeAccessTokenOnSignout: true,
        metadata: {
            issuer: "https://betting-game.eu.auth0.com/",
            authorization_endpoint:
                "https://betting-game.eu.auth0.com/authorize",
            userinfo_endpoint: "https://betting-game.eu.auth0.com/userinfo",
            token_endpoint: "https://betting-game.eu.auth0.com/oauth/token",
            registration_endpoint:
                "https://betting-game.eu.auth0.com/oidc/register",
            revocation_endpoint:
                "https://betting-game.eu.auth0.com/oauth/revoke",
            end_session_endpoint:
                "https://betting-game.eu.auth0.com/v2/logout?client_id=rqlyAphr2es3RUjYlPAEkf2VCldAR5wL&returnTo=http://localhost:3000/",
            jwks_uri: "https://betting-game.eu.auth0.com/",
        },
    });
    return <>{children}</>;
};

export default AuthContext;
