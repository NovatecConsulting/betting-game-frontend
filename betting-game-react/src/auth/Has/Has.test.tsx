import React from "react";
import { mount } from "enzyme";
import { Has } from ".";
import { AuthorizationManager } from "../../auth";
import jwt_decode from "jwt-decode";

jest.mock("../../auth", () => ({
    AuthorizationManager: {
        getUser: jest.fn(),
    },
}));

jest.mock("jwt-decode");

describe(Has.name + " component", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("should render and match snapshot, if user has all the needed permission.", async () => {
        const userMock = { access_token: "testtoken" } as any;
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [true, hasAccessMock];

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            Promise.resolve(userMock)
        );
        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: ["TestPermission1", "TestPermission2"],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        expect(hasAccessMock).toHaveBeenCalledWith(true);
        expect(wrapper).toIncludeText("AccessGranted");
        expect(wrapper).not.toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if user has more than the needed permissions.", async () => {
        const userMock = { access_token: "testtoken" } as any;
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [true, hasAccessMock];

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            Promise.resolve(userMock)
        );
        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: ["TestPermission1", "TestPermission2"],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        expect(hasAccessMock).toHaveBeenCalledWith(true);
        expect(wrapper).toIncludeText("AccessGranted");
        expect(wrapper).not.toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if user has just one of the needed permissions.", async () => {
        const userMock = { access_token: "testtoken" } as any;
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [false, hasAccessMock];

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            Promise.resolve(userMock)
        );
        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: ["TestPermission1"],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        expect(hasAccessMock).toHaveBeenCalledWith(false);
        expect(wrapper).not.toIncludeText("AccessGranted");
        expect(wrapper).toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if user has no permissions.", async () => {
        const userMock = { access_token: "testtoken" } as any;
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [false, hasAccessMock];

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            Promise.resolve(userMock)
        );
        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: [],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        expect(hasAccessMock).toHaveBeenCalledWith(false);
        expect(wrapper).not.toIncludeText("AccessGranted");
        expect(wrapper).toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if user has no access token.", async () => {
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [false, hasAccessMock];
        const userMock = {} as any;
        const mockPromise = Promise.resolve(userMock);

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            mockPromise
        );

        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: ["TestPermission1"],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        await mockPromise; // Wait for promise

        expect(hasAccessMock).toHaveBeenCalledWith(false);
        expect(wrapper).not.toIncludeText("AccessGranted");
        expect(wrapper).toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if jwt_decode throws an error.", async () => {
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [false, hasAccessMock];
        const userMock = { access_token: "testtoken" } as any;
        const mockPromise = Promise.resolve(userMock);

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            mockPromise
        );

        (jwt_decode as jest.Mock).mockImplementation(() => {
            throw new Error();
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        await mockPromise; // Wait for promise

        expect(hasAccessMock).toHaveBeenCalledWith(false);
        expect(wrapper).not.toIncludeText("AccessGranted");
        expect(wrapper).toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, if AuthorizationManager throws an error.", async () => {
        const hasAccessMock = jest.fn();
        const useStateMock: any = () => [false, hasAccessMock];
        const mockPromise = Promise.reject("TestError");

        jest.spyOn(React, "useState").mockImplementationOnce(useStateMock);
        (AuthorizationManager.getUser as jest.Mock).mockReturnValueOnce(
            mockPromise
        );
        (jwt_decode as jest.Mock).mockReturnValueOnce({
            permissions: ["TestPermission1"],
        });

        const wrapper = await mount(
            <Has
                permissions={["TestPermission1", "TestPermission2"]}
                Yes={() => <div>AccessGranted</div>}
                No={() => <div>AccessDenied</div>}
            />
        );

        await mockPromise.catch(() => {}); // Wait for promise

        expect(hasAccessMock).toHaveBeenCalledWith(false);
        expect(wrapper).not.toIncludeText("AccessGranted");
        expect(wrapper).toIncludeText("AccessDenied");
        expect(wrapper).toMatchSnapshot();
    });
});
