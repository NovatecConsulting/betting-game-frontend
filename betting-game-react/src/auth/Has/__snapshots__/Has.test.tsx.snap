// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Has component should render and match snapshot, if AuthorizationManager throws an error. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <No>
    <div>
      AccessDenied
    </div>
  </No>
</Has>
`;

exports[`Has component should render and match snapshot, if jwt_decode throws an error. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <No>
    <div>
      AccessDenied
    </div>
  </No>
</Has>
`;

exports[`Has component should render and match snapshot, if user has all the needed permission. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <Yes>
    <div>
      AccessGranted
    </div>
  </Yes>
</Has>
`;

exports[`Has component should render and match snapshot, if user has just one of the needed permissions. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <No>
    <div>
      AccessDenied
    </div>
  </No>
</Has>
`;

exports[`Has component should render and match snapshot, if user has more than the needed permissions. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
    ]
  }
>
  <Yes>
    <div>
      AccessGranted
    </div>
  </Yes>
</Has>
`;

exports[`Has component should render and match snapshot, if user has no access token. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <No>
    <div>
      AccessDenied
    </div>
  </No>
</Has>
`;

exports[`Has component should render and match snapshot, if user has no permissions. 1`] = `
<Has
  No={[Function]}
  Yes={[Function]}
  permissions={
    Array [
      "TestPermission1",
      "TestPermission2",
    ]
  }
>
  <No>
    <div>
      AccessDenied
    </div>
  </No>
</Has>
`;
